### DISCUSSION TITLE
Is there a way to ask for credentials during add-on installation?

### DISCUSSION THREAD

**Discussion by ygerasimov** [docksal/docksal#1830] [Q&A]
I would like to build an addon for Diffy (https://diffy.website). We need to collect API key and project ID. Is there an example of best practices how to accomplish that during installation? Ideally to save them as environment variables but if that is not an option I can save them to an `.env` file.

**Comment by lmakarov** [comment]
Docksal supports project-level and global configuration variables - https://docs.docksal.io/stack/configuration-variables/.

In your case, project ID would be set as a project-level variable, while the API key can be either project level (stored in the project repo and shared) or global (individual per machine):

```bash
# Set a project scope variable
$ fin config set DIFFY_PROJECT_ID=<...>

# Set a global scope variable
$ fin config set --global SECRET_DIFFY_API_KEY=<...>
```

Prefixing variables with `SECRET_` ensures that their values are masked in output of `fin` commands (e.g., `fin config` or `fin diagnose`). Project level variables will override the global ones if their names overlap.

To prompt for these during the add-on installation, you can use something like this:

```bash
# Prompt for Diffy project ID
while [[ -z "${diffy_project_id}" ]]; do
    read -p "Diffy project ID: " diffy_project_id
    [[ -z "${diffy_project_id}" ]] && echo "Diffy project ID cannot be empty!"
done
# Store the Diffy project ID as a project-level variable
fin config set DIFFY_PROJECT_ID=${diffy_project_id}

# Prompt for Diffy API key
while [[ -z "${diffy_api_key}" ]]; do
    read -s -p "Diffy API key: " diffy_api_key
    [[ -z "${diffy_api_key}" ]] && echo "Diffy API key cannot be empty!"
done
# Store the Diffy API key as a global variable
fin config set --global SECRET_DIFFY_API_KEY=${diffy_api_key}

```
